Assets/Scripts/SS3D/Systems/IngameConsoleSystem/CommandsController.cs

```diff
@@ -29,33 +40,83 @@ public CommandsController()
         }
+
+        [Client]
+        public void ClientProcessCommand(string command)
+        {
+            CmdProcessCommand(command);
+        }
+
         /// <summary>
         /// Find and call command
         /// </summary>
         /// <param name="command">Command and it's args separated by spaces</param>
         /// <returns>Command response</returns>
-        public string ProcessCommand(string command)
+        [ServerRpc(RequireOwnership = false)]
+        public void CmdProcessCommand(string command, NetworkConnection conn = null)
         {
+            string ckey = Subsystems.Get<PlayerSystem>().GetCkey(conn);
+            if (!Subsystems.Get<PermissionSystem>().TryGetUserRole(ckey, out ServerRoleTypes userPermission))
+            {
+                RpcCommandAnswer(conn, string.Format("No role found for user {0}, can't process command", ckey));
+                return;
+            }
+
             string[] splitCommand = command.Split(' ');
-            string commandName = splitCommand[0]; 
+            string commandName = splitCommand[0];
             if (commandName == "help")
             {
-                return HelpCommand();
+                RpcCommandAnswer(conn, HelpCommand());
+                return;
             }
             if (splitCommand.Length > 1)
             {
                 if (splitCommand[1] == "help")
                 {
-                    return LongHelpCommand(command);
+                    RpcCommandAnswer(conn, LongHelpCommand(command));
+                    return;
                 }
             }
-            // Find proper command body and prepare correct args
-            if (_allCommands.ContainsKey(commandName))
+
+            if (!_allCommands.ContainsKey(commandName))
             {
-                return _allCommands[commandName].Perform(command.Split().Skip(1).ToArray());
-            } 
-            return "No such command exists";
+                RpcCommandAnswer(conn, "No such command exists");
+                return;
+            }
+
+            if (_allCommands[commandName].AccessLevel > userPermission)
+            {
+                RpcCommandAnswer(conn, "Access level too low, can't perform command");
+                return;
+            }
+
+            if (_allCommands[commandName].ServerCommand)
+            {
+                RpcCommandAnswer(conn, _allCommands[commandName].Perform(command.Split().Skip(1).ToArray(), conn));
+                return;
+            }
+            else
+            {
+                RpcPerformOnClient(conn, command);
+                return;
+            }
         }
+
+        [TargetRpc]
+        public void RpcPerformOnClient(NetworkConnection conn, string command)
+        {
+            string[] splitCommand = command.Split(' ');
+            string commandName = splitCommand[0];
+            string answer = _allCommands[commandName].Perform(command.Split().Skip(1).ToArray());
+            console.AddText(answer);
```

Comment:
```suggestion
        string[] splitCommand = command.Split(' ');
        if (splitCommand.Length <= 0)
        {
            console.AddText("Command is empty");
            return;
        }
        string commandName = splitCommand[0];
        if (!_allCommands.TryGetValue(commandName, out var commandHandler))
        {
            console.AddText("No such command exists");
            return;
        }
        string[] arguments = splitCommand.Skip(1).ToArray();
        string answer = commandHandler.Perform(arguments);
        console.AddText(answer);
```

Command is missing necessary safety checks